ClassCastException from FunctionDeclaration to MethodDeclaration (Java)

The file causing this crash is [GraphZoomer.java from google bindiff](https://github.com/google/bindiff/blob/main/java/ui/src/main/java/com/google/security/zynamics/bindiff/graph/helpers/GraphZoomer.java):

The line that causes the crash is: `SwingUtilities.invokeLater(() -> zoomToNodes((BinDiffGraph<ZyGraphNode<?>, ?>) graph, nodes));`

Commenting out this line prevents the CPG from crashing, maybe there is an issue with the lambda notation?

# Stack trace

```
09:48:50,717 INFO  TranslationConfiguration$Builder Registered language frontend 'CLanguage' for following file types: [c, h]
09:48:50,725 INFO  TranslationConfiguration$Builder Registered language frontend 'CPPLanguage' for following file types: [cpp, cc, cxx, hpp, hh]
09:48:50,732 INFO  TranslationConfiguration$Builder Registered language frontend 'JavaLanguage' for following file types: [java]
09:48:50,733 INFO  TranslationConfiguration$Builder Registered language frontend 'JavaLanguage' for following file types: [java]
09:48:50,737 INFO  TranslationConfiguration$Builder Registered language frontend 'GoLanguage' for following file types: [go]
09:48:50,756 INFO  TranslationConfiguration$Builder Registered an extra (frontend dependent) default dependency: class de.fraunhofer.aisec.cpg.passes.FunctionPointerCallResolver
09:48:50,756 INFO  TranslationConfiguration$Builder Registered an extra (frontend dependent) default dependency: class de.fraunhofer.aisec.cpg.passes.FunctionPointerCallResolver
09:48:50,758 INFO  TranslationConfiguration$Builder Registered an extra (frontend dependent) default dependency: class de.fraunhofer.aisec.cpg.passes.JavaExternalTypeHierarchyResolver
09:48:50,759 INFO  TranslationConfiguration$Builder Registered an extra (frontend dependent) default dependency: class de.fraunhofer.aisec.cpg.passes.JavaExternalTypeHierarchyResolver
09:48:50,763 INFO  TranslationConfiguration$Builder Registered an extra (frontend dependent) default dependency: class de.fraunhofer.aisec.cpg.passes.GoExtraPass
09:48:50,765 INFO  TranslationConfiguration$Builder Passes before enforcing order: [TypeHierarchyResolver, ImportResolver, VariableUsageResolver, CallResolver, DFGPass, EvaluationOrderGraphPass, TypeResolver, ControlFlowSensitiveDFGPass, FilenameMapper, PrepareSerialization, FunctionPointerCallResolver, FunctionPointerCallResolver, JavaExternalTypeHierarchyResolver, JavaExternalTypeHierarchyResolver, GoExtraPass]
09:48:50,845 INFO  TranslationConfiguration$Builder Passes after enforcing order: [TypeHierarchyResolver, PrepareSerialization, JavaExternalTypeHierarchyResolver, ImportResolver, GoExtraPass, VariableUsageResolver, CallResolver, DFGPass, EvaluationOrderGraphPass, TypeResolver, ControlFlowSensitiveDFGPass, FunctionPointerCallResolver, FilenameMapper]
09:48:50,848 INFO  TranslationConfiguration$Builder Passes before enforcing order: []
09:48:50,848 INFO  TranslationConfiguration$Builder Passes after enforcing order: []
09:48:50,870 INFO  MeasurementHolder TranslationManager: Translation into full graph
09:48:50,872 INFO  MeasurementHolder TranslationManager: Executing Language Frontend
09:48:50,879 INFO  TranslationManager Parsing /home/aisec/rh/cpg-test/test-projects/java/bindiff-8/java/ui/src/main/java/com/google/security/zynamics/bindiff/graph/helpers/GraphZoomer.java
09:48:51,014 INFO  JavaLanguageFrontend Source file root used for type solver: /home/aisec/rh/cpg-test/test-projects/java/bindiff-8/java/ui/src/main/java/com/google/security/zynamics/bindiff/graph/helpers
09:48:51,092 INFO  MeasurementHolder JavaLanguageFrontend: Parsing source file
09:48:51,271 INFO  MeasurementHolder JavaLanguageFrontend: Parsing source file done in 172 ms
09:48:51,271 INFO  MeasurementHolder JavaLanguageFrontend: Transform to CPG
09:48:52,111 INFO  MeasurementHolder TranslationManager: Translation into full graph done in 1239 ms
java.util.concurrent.ExecutionException: java.lang.ClassCastException: class de.fraunhofer.aisec.cpg.graph.declarations.FunctionDeclaration cannot be cast to class de.fraunhofer.aisec.cpg.graph.declarations.MethodDeclaration (de.fraunhofer.aisec.cpg.graph.declarations.FunctionDeclaration and de.fraunhofer.aisec.cpg.graph.declarations.MethodDeclaration are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at de.fraunhofer.aisec.cpg_vis_neo4j.Application.call(Application.kt:443)
	at de.fraunhofer.aisec.cpg_vis_neo4j.Application.call(Application.kt:69)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at de.fraunhofer.aisec.cpg_vis_neo4j.ApplicationKt.main(Application.kt:483)
Caused by: java.lang.ClassCastException: class de.fraunhofer.aisec.cpg.graph.declarations.FunctionDeclaration cannot be cast to class de.fraunhofer.aisec.cpg.graph.declarations.MethodDeclaration (de.fraunhofer.aisec.cpg.graph.declarations.FunctionDeclaration and de.fraunhofer.aisec.cpg.graph.declarations.MethodDeclaration are in unnamed module of loader 'app')
	at de.fraunhofer.aisec.cpg.frontends.java.ExpressionHandler.createImplicitThis(ExpressionHandler.kt:798)
	at de.fraunhofer.aisec.cpg.frontends.java.ExpressionHandler.handleMethodCallExpression(ExpressionHandler.kt:763)
	at de.fraunhofer.aisec.cpg.frontends.java.ExpressionHandler._init_$lambda$29(ExpressionHandler.kt:902)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.kt:107)
	at de.fraunhofer.aisec.cpg.frontends.java.StatementHandler.handleExpressionStatement(StatementHandler.kt:63)
	at de.fraunhofer.aisec.cpg.frontends.java.StatementHandler._init_$lambda$35(StatementHandler.kt:636)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.kt:107)
	at de.fraunhofer.aisec.cpg.frontends.java.ExpressionHandler.handleLambdaExpr(ExpressionHandler.kt:73)
	at de.fraunhofer.aisec.cpg.frontends.java.ExpressionHandler._init_$lambda$37(ExpressionHandler.kt:910)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.kt:107)
	at de.fraunhofer.aisec.cpg.frontends.java.ExpressionHandler.handleMethodCallExpression(ExpressionHandler.kt:779)
	at de.fraunhofer.aisec.cpg.frontends.java.ExpressionHandler._init_$lambda$29(ExpressionHandler.kt:902)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.kt:107)
	at de.fraunhofer.aisec.cpg.frontends.java.StatementHandler.handleExpressionStatement(StatementHandler.kt:63)
	at de.fraunhofer.aisec.cpg.frontends.java.StatementHandler._init_$lambda$35(StatementHandler.kt:636)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.kt:107)
	at de.fraunhofer.aisec.cpg.frontends.java.StatementHandler.handleBlockStatement(StatementHandler.kt:344)
	at de.fraunhofer.aisec.cpg.frontends.java.StatementHandler._init_$lambda$32(StatementHandler.kt:627)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.kt:107)
	at de.fraunhofer.aisec.cpg.frontends.java.DeclarationHandler.handleMethodDeclaration(DeclarationHandler.kt:159)
	at de.fraunhofer.aisec.cpg.frontends.java.DeclarationHandler._init_$lambda$18(DeclarationHandler.kt:435)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.kt:107)
	at de.fraunhofer.aisec.cpg.frontends.java.DeclarationHandler.handleClassOrInterfaceDeclaration(DeclarationHandler.kt:241)
	at de.fraunhofer.aisec.cpg.frontends.java.DeclarationHandler._init_$lambda$20(DeclarationHandler.kt:442)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.kt:107)
	at de.fraunhofer.aisec.cpg.frontends.java.JavaLanguageFrontend.parse(JavaLanguageFrontend.kt:128)
	at de.fraunhofer.aisec.cpg.TranslationManager.parse(TranslationManager.kt:363)
	at de.fraunhofer.aisec.cpg.TranslationManager.parseSequentially(TranslationManager.kt:319)
	at de.fraunhofer.aisec.cpg.TranslationManager.runFrontends(TranslationManager.kt:229)
	at de.fraunhofer.aisec.cpg.TranslationManager.analyzeNonAsync(TranslationManager.kt:92)
	at de.fraunhofer.aisec.cpg.TranslationManager.analyze$lambda$0(TranslationManager.kt:74)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)

```

# Code of interest


```
...

public class GraphZoomer {
  ...

  @SuppressWarnings("unchecked")
  public static void zoomToNode(final BinDiffGraph<?, ?> graph, final ZyGraphNode<?> node) {
    checkNotNull(graph);
    checkNotNull(node);

    if (!node.isVisible()) {
      GraphLayoutEventHandler.handleUnhideInvisibleNode(
          (BinDiffGraph<ZyGraphNode<?>, ?>) graph, node);
    }

    final Collection<ZyGraphNode<?>> nodes = new ArrayList<>();
    nodes.add(node);

    SwingUtilities.invokeLater(() -> zoomToNodes((BinDiffGraph<ZyGraphNode<?>, ?>) graph, nodes));
  }

  ...
}

```

--- a/cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/ExpressionHandler.kt
+++ b/cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/ExpressionHandler.kt
@@ -792,7 +792,7 @@ class ExpressionHandler(lang: JavaLanguageFrontend) :
         de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression {
         val base: de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression
         val thisType =
-            (frontend.scopeManager.currentFunction as MethodDeclaration?)?.receiver?.type
+            (frontend.scopeManager.currentFunction as? MethodDeclaration?)?.receiver?.type
                 ?: unknownType()
         base = newReference("this", thisType).implicit("this")
         return base

